### PR TITLE
v3.1.0

### PR DESCRIPTION
**Pull request by fed135**
## Summary
v3.1 proposal

## Changelist

- Rewrote the queries system to be more maintainable
- Replaced the storage library to a full js one, enhancing portability of this library
- Reduced memory usage by ~4x
- Added the `cause` attribute to the query event ("limit" or "timeout")

## Breaking changes

- Changed the output of the `size` method
- Changed the payload format of `cacheHit`, `cacheMiss` and `coalescedHit` to directly output the number

## Merge Checklist

 - [x] Verified locally
 - [ ] Unit Tests added
 - [ ] Integration tests added
 - [x] Unit tests pass locally
 - [x] Integration tests pass locally

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,14 +77,14 @@ batch | false | <pre>{&#13;&#10;&nbsp;&nbsp;delay: 50,&#13;&#10;&nbsp;&nbsp;limi
 
 HA-store emits events to track cache hits, miss and outbound requests.
 
-Event | Description
+Event | Format | Description
 --- | ---
```
Comment (mats852) [inline README.md:81]: [⛰️] Table is broken, this fixes it
```suggestion
--- | --- | ---
```

File: src/buffer.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      this.ids = [];
+      this.contextKey = key;
+      this.params = params;
+      this.contexts = [];
+      this.handle = deferred();
+      this.timer = null;
+    }
+
+    tick() {
+      const sizeLimit = config.batch?.limit || 1;
```
Comment (mats852) [inline src/buffer.js:23]: [🧗] Do you think evaluating those values only once could give a performance benefit?

File: src/buffer.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+      if (this.ids.length >= sizeLimit) {
+        this.run('limit');
+        return this;
+      }
+      if (this.timer === null) {
+        this.timer = setTimeout(this.run.bind(this, 'timeout'), config.batch?.delay || 0);
```
Comment (mats852) [inline src/buffer.js:30]: Same as above

File: src/buffer.js (hunk 2 above)
Comment (mats852) [reply src/buffer.js:23]: If it's microscopic please disregard, let's not impair the readability of the code for that.

File: src/buffer.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      this.timer = null;
+    }
+
+    tick() {
+      const sizeLimit = config.batch?.limit || 1;
+
+      if (this.ids.length >= sizeLimit) {
+        this.run('limit');
+        return this;
+      }
```
Comment (mats852) [inline src/buffer.js:28]: [💨] Line return 😅 
```suggestion
      }
      
```

File: src/buffer.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  const buffers = [];
+
+  let numCoalesced = 0;
+  let numCached = 0;
+  let numMisses = 0;
+
+  class RequestBuffer {
+    constructor(key, params) {
+      this.uid = Math.random().toString(36);
+      this.state = 0;
```
Comment (mats852) [inline src/buffer.js:13]: [⚪] What does this map to? Any chance we can replace those magic values for an "enum" ?

File: src/buffer.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    handleQuerySuccess(entries) {
+      this.state = 2;
+      emitter.emit('querySuccess', { key: this.contextKey, uid: this.uid, size: this.ids.length, params: this.params });
+      this.handle.resolve(entries);
+      if (config.cache !== null) targetStore.set(contextRecordKey(this.contextKey), this.ids, entries || {});
+      buffers.splice(buffers.indexOf(this), 1);
+    }
+  }
+
+  const getHandles = (async function(key, ids, params, context) {
```
Comment (mats852) [inline src/buffer.js:73]: Just curious why we're not using this form here

```js
   async function getHandles(key, ids, params, context) {
```

File: src/buffer.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    for (let i = 0; i < ids.length; i++) {
+      if (handles[i] === undefined) {
+        const liveBuffer = buffers.find(buffer => buffer.contextKey === key && buffer.ids.includes(ids[i]));
+        if (liveBuffer) {
+          numCoalesced++;
+          handles[i] = liveBuffer.handle.promise.then(results => results[ids[i]]);
+        }
+        else {
```
Comment (mats852) [inline src/buffer.js:93]: [⚪] Just to make sure if it's the code style or not
```suggestion
        } else {
```

File: src/buffer.js (hunk 2 above)
Comment (fed135) [reply src/buffer.js:23]: Indeed, not a huge gain.

File: src/buffer.js (hunk 3 above)
Comment (fed135) [reply src/buffer.js:30]: The bind here is probably more expensive, but that's the price to pay to use Classes in JS -.-

File: src/buffer.js (hunk 5 above)
Comment (fed135) [reply src/buffer.js:13]: Good point, I'll make one

File: src/buffer.js (hunk 6 above)
Comment (fed135) [reply src/buffer.js]: Corrected- I was even able to remove the async flag 🎉

File: src/buffer.js (hunk 7 above)
Comment (fed135) [reply src/buffer.js:87]: 🤷  Linter passed

Will check rest of codebase

File: src/buffer.js (hunk 5 above)
Comment (fed135) [reply src/buffer.js]: @mats852 After switching to an enum, memory usage dropped by 4x ! 
Now that's an interesting case study!

Comment (mats852) [reply src/buffer.js]: Never expected this! In hindsight it can make sense. Great!!

File: src/buffer.js (hunk 3 above)
Comment (mats852) [reply src/buffer.js:37]: I guess if this was of paramount importance we could set these parameters once in the constructor and just call `this.config.whathever`!

Comment (mats852) [reply src/buffer.js:37]: I'll still approve :)

**Pull request merged**